File: CodingChallengeSolutionsC++/2145. Count the Hidden Sequences/2145.CounttheHiddenSequences.py
class CounttheHiddenSequences:
    def numberOfArrays(self, differences: List[int], lower: int, upper: int) -> int:
        presum = 0
        minval = 0
        maxval = 0

        for x in differences:
            presum += x
            minval = min(minval, presum)
            maxval = max(maxval, presum)
    
        return max(0, (upper-lower+1) - (maxval - minval))

File: CodingChallengeSolutionsC++/2145. Count the Hidden Sequences/2145.CounttheHiddenSequences.cpp
#include <vector>
class CounttheHiddenSequences {
public:
    int numberOfArrays(std::vector<int>& differences, int lower, int upper) {
        // // Brute Force (O(n^2))
        // int result = 0;

        // for(int x = lower; x <= upper; ++x)
        // {
        //     // Start with the lower bound
        //     int preSum = x;
        //     for(int y = 0; y < differences.size(); ++y)
        //     {
        //         preSum += differences[y];
        //         if(preSum < lower || preSum > upper)
        //         {
        //             break;
        //         }
        //         else if(y == differences.size()-1)
        //         {
        //             ++result;
        //         }
        //     } 
        // }
        // return result;

        // 
        long preSum = 0;
        long minVal = 0;
        long maxVal = 0;
        for (int x = 0; x < differences.size(); ++x)
        {
            preSum += differences[x];
            minVal = min(minVal, preSum);
            maxVal = max(maxVal, preSum);
        }

        return max(0L, (upper - lower + 1) - (maxVal - minVal));
    }
};